### PR TITLE
feat: add the workflow to release the installing package for macos

### LINKED ISSUE #1894: Auto release the installing package of macos
add the workflow to release the installing package of macos  automatically when a new tag created

### PR DESCRIPTION
**Pull request by mangoGoForward**
Signed-off-by: mango <xu.weiKyrie@foxmail.com>

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

* **What is the current behavior?** (You can also link to an open issue here)
Fixed #1894

* **What is the new behavior (if this is a feature change)?**
None

### CODE REVIEW

File: .github/workflows/cicd.yaml [hunk 1]
```diff
@@ -141,6 +141,93 @@ jobs:
           check_name: Linux Test Report
           comment_title: Linux Test Report
 
+  macos-cpp:
+    runs-on: macos-latest
+    if: github.repository == '4paradigm/OpenMLDB'
+    container:
+      image: ghcr.io/4paradigm/hybridsql:latest
```
Comment (aceforeverd) [inline .github/workflows/cicd.yaml]: macos job do not runs on container

File: .github/workflows/cicd.yaml [hunk 2]
```diff
@@ -141,6 +141,93 @@ jobs:
           check_name: Linux Test Report
           comment_title: Linux Test Report
 
+  macos-cpp:
+    runs-on: macos-latest
+    if: github.repository == '4paradigm/OpenMLDB'
```
Comment (aceforeverd) [inline .github/workflows/cicd.yaml]: only runs it if is a tag push

File: .github/workflows/cicd.yaml [hunk 3]
```diff
@@ -141,6 +141,93 @@ jobs:
...
+    container:
+      image: ghcr.io/4paradigm/hybridsql:latest
+    env:
+      OS: darwin
+      ARCH: x86_64
+      NPROC: 3
+      CTEST_PARALLEL_LEVEL: 3 # parallel test level for ctest (make test)
```
Comment (aceforeverd) [inline .github/workflows/cicd.yaml]: set to 1, parallel runs will fail

File: .github/workflows/cicd.yaml [hunk 4]
```diff
@@ -141,6 +141,93 @@ jobs:
...
+
+      - name: build
+        run: |
+          make build
+          # GitHub runner disk space is limited
+          # delete thirdparty build directory($ROOT/.deps/) to save disk space
+          # make thirdpartybuild-clean
+
+      - name: run cpp
+        id: unittest
```
Comment (aceforeverd) [inline .github/workflows/cicd.yaml]: I guess we can omit unit tests, as well as the tests below @dl239

File: .github/workflows/cicd.yaml [hunk 5]
```diff
@@ -141,6 +141,93 @@ jobs:
...
+        run: |
+          tar czf ${{ env.OPENMLDB_PREFIX }}.tar.gz ${{ env.OPENMLDB_PREFIX }}/
+
+      - name: upload artifacts
+        if: ${{ github.event_name == 'push' }}
+        uses: actions/upload-artifact@v2
+        with:
+          path: openmldb-*.tar.gz
+          name: release-artifacts
+
```
Comment (aceforeverd) [inline .github/workflows/cicd.yaml:199]: release dependencies need update https://github.com/4paradigm/OpenMLDB/pull/2028/files#diff-2c9df400afaffa9270cab617792f08cc671e5f95ae192f1cf111ef4e990a4addR234

File: .github/workflows/cicd.yaml (hunk 1 above)
Comment (mangoGoForward) [reply .github/workflows/cicd.yaml]: Done

File: .github/workflows/cicd.yaml (hunk 2 above)
Comment (mangoGoForward) [reply .github/workflows/cicd.yaml]: Done

File: .github/workflows/cicd.yaml (hunk 3 above)
Comment (mangoGoForward) [reply .github/workflows/cicd.yaml]: Done

File: .github/workflows/cicd.yaml (hunk 4 above)
Comment (mangoGoForward) [reply .github/workflows/cicd.yaml]: Done

Comment (mangoGoForward) [reply .github/workflows/cicd.yaml]: Sorry, some steps deleted by mistake, I will fix it

Comment (dl239) [reply .github/workflows/cicd.yaml]: yes, no need to run tests

File: .github/workflows/cicd.yaml (hunk 5 above)
Comment (mangoGoForward) [reply .github/workflows/cicd.yaml:199]: I do not really understand, could you please give me some advice?

Comment (aceforeverd) [reply .github/workflows/cicd.yaml:199]: So basically the release flow is `tag push` -> `trigger this cicd workflow and runs cpp and cpp-macos` -> `runs the release job below` , and finally you can find the new release in https://github.com/4paradigm/OpenMLDB/releases. 

This step uploaded the macos artifact just built here, you can also find the same step in `cpp` job. And the `release` job just download linux and macos artifacts and release them once for all

This graph may help https://github.com/4paradigm/OpenMLDB/actions/runs/2468485230

Comment (mangoGoForward) [reply .github/workflows/cicd.yaml:199]: > So basically the release flow is `tag push` -> `trigger this cicd workflow and runs cpp and cpp-macos` -> `runs the release job below` , and finally you can find the new release in https://github.com/4paradigm/OpenMLDB/releases.
> 
> This step uploaded the macos artifact just built here, you can also find the same step in `cpp` job. And the `release` job just download linux and macos artifacts and release them once for all


This flow seems is consistent with you described~, whether the `runs-on` in `release` step will affect the action of release linux and macos artifacts?

Comment (mangoGoForward) [reply .github/workflows/cicd.yaml:199]: > > So basically the release flow is `tag push` -> `trigger this cicd workflow and runs cpp and cpp-macos` -> `runs the release job below` , and finally you can find the new release in https://github.com/4paradigm/OpenMLDB/releases.
> > This step uploaded the macos artifact just built here, you can also find the same step in `cpp` job. And the `release` job just download linux and macos artifacts and release them once for all
> 

This flow seems is consistent with you described~, whether the `runs-on` in `release` step will affect the action of release linux and macos artifacts?

https://github.com/4paradigm/OpenMLDB/blob/af0272cec5b9646c8b7eaa3c6d9e9dfccaf6f56d/.github/workflows/cicd.yaml#L144-L145

Comment (aceforeverd) [reply .github/workflows/cicd.yaml:199]: the release job just publish the built artifacts of needed two job, so it do not matter which platform release job runs on. 

Built from multiple platform and release them once for all

**Comment by aceforeverd** [comment]
Github workflow fails itself 🥲

File: .github/workflows/cicd.yaml (hunk 5 above)
Comment (mangoGoForward) [reply .github/workflows/cicd.yaml:199]: Got, thanks.

**Comment by mangoGoForward** [comment]
> Github workflow fails itself 🥲

Too strange, and I was confused why this pr has no checks yesterday

**Comment by mangoGoForward** [comment]
@aceforeverd The CI of `macos-cpp` and `macos-release` has been skipped, because only push a tag can trigger the flow.
but we can see the graph, it is seems right?
 https://github.com/4paradigm/OpenMLDB/actions/runs/2554887555

image

**Comment by aceforeverd** [comment]
> @aceforeverd The CI of `macos-cpp` and `macos-release` has been skipped, because only push a tag can trigger the flow. but we can see the graph, it is seems right? https://github.com/4paradigm/OpenMLDB/actions/runs/2554887555
> 
> image

yes, it is the expected behavior for pull request. No worry I will test the release logic later today

**Comment by mangoGoForward** [comment]
> yes, it is the expected behavior for pull request. No worry I will test the release logic later today

OK

File: .github/workflows/cicd.yaml [hunk 6]
```diff
@@ -160,3 +216,23 @@ jobs:
             openmldb-*.tar.gz
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  macos-release:
```
Comment (aceforeverd) [inline .github/workflows/cicd.yaml]: foroget to see this. there is no need to add a new `macos-release` job here. 
the only thing needed for release job is update dependency in line https://github.com/4paradigm/OpenMLDB/pull/2028/files#diff-2c9df400afaffa9270cab617792f08cc671e5f95ae192f1cf111ef4e990a4addR203. 

someting like this:
```
needs: [ "cpp", "macos-cpp" ]
```

Comment (mangoGoForward) [reply .github/workflows/cicd.yaml]: Done. Thank you for your patient guidance

**Review by aceforeverd — APPROVED** [review]
Tested and goes well. Thanks for being patient !

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Pull request merged**